Add translation of 3-additional-articles/9-regular-expressions/11-regexp-groups (#706)
javascript-tutorial/zh.javascript.info: 9-regular-expressions/11-regexp-groups/article.md

# 捕获组

> 模式的一部分可以用括号括起来 `pattern:(...)`。这称为“捕获组”。

Reviewer: ```suggestion
模式的一部分可以用括号括起来 `pattern:(...)`。这称为“捕获组（capturing group）”。
```

Reviewer: 这里是因为首次出现，要标注一下英文。

alert( tag[1] ); // h1
```

### 嵌套组

括号可以嵌套。在这种情况下，编号也从左到右。

例如，在搜索标签 `subject:<span class="my">` 时我们可能会对以下内容感兴趣：

1. 整个标签内容：`match:span class="my"`。
2. 标签名称：`match:span`。
3. 标签属性：`match:class="my"`。

让我们为它们添加括号：`pattern:<(([a-z]+)\s*([^>]*))>`。

这是它们的编号方式（从左到左，由父模式开始）：

> ![](regexp-nested-groups-pattern.svg)

Reviewer 2: 麻烦检查一下这里的图片，好像没有显示？

![image](https://user-images.githubusercontent.com/45781328/78556146-04629b80-7841-11ea-9aa2-3a4167e5a096.png)

参考英文：
![image](https://user-images.githubusercontent.com/45781328/78556311-5acfda00-7841-11ea-9187-2025ac91bccb.png)

Author: 我这里没问题，你可以试试最新版的chrome

alert(result[3]); // class="my"
```

`result` 的零索引始终保持完全匹配。

然后按左括号将组从左到右编号。第一组返回为 `result[1]`。它包含了整个标签内容。

然后 `result[2]` 从第二个开始的括号中进入该组 `pattern:([a-z]+)` —— 标签名称，然后在 `result[3]` 标签中：`pattern:([^>]*)`。

字符串中每个组的内容：

> ![](regexp-nested-groups-matches.svg)

Reviewer 2: 同样这个图片也没显示

Author: 也没问题

### 可选组

即使组是可选的并且在匹配项中不存在（例如，具有数量词 `pattern:(...)?`），也存在相应的 `result` 数组项，并且等于 `undefined`。

例如，让我们考虑正则 `pattern:a(z)?(c)?`。它寻找 `"a"` ，然后是可选的 `"z"`，然后是可选的 `"c"`。

如果我们在单个字母的字符串上运行 `subject:a`，则结果为：

```js run
let match = 'a'.match(/a(z)?(c)?/);

alert( match.length ); // 3
alert( match[0] ); // a（完全匹配）
alert( match[1] ); // undefined
alert( match[2] ); // undefined
```

数组的长度为 `3`，但所有组均为空。

这是字符串的一个更复杂的匹配 `subject:ac`：

```js run
let match = 'ac'.match(/a(z)?(c)?/)

alert( match.length ); // 3
alert( match[0] ); // ac（完全匹配）

> alert( match[1] ); // undefined, 因为 (z)? 没匹配项

Reviewer: ```suggestion
alert( match[1] ); // undefined，因为 (z)? 没匹配项
```

alert( tags ); // <h1>,<h2>
```

结果是一个匹配数组，但没有每个匹配项的详细信息。但是实际上，我们通常需要在结果中获取捕获组的内容。

要获取它们，我们应该使用方法 `str.matchAll(regexp)` 进行搜索。

> 在使用 `match` 很长一段时间后，它作为“新的改进版本”被加入到 JavaScript 中

Reviewer 2: ```suggestion
在使用 `match` 很长一段时间后，它作为“新的改进版本”被加入到 JavaScript 中。
```

> 就像 `match` 一样，它寻找匹配项，但有3个区别：

Reviewer 2: ```suggestion
就像 `match` 一样，它寻找匹配项，但有 3 个区别：
```

> 1. 它不返回数组，而是一个可迭代的对象。

Reviewer 2: ```suggestion
1. 它返回的不是数组，而是一个可迭代的对象。
```

alert(results[1]); // <h2>,h2 (2nd tag)
```

> 我们可以看到，第一个区别非常重要，如 `(*)` 行所示。我们无法获得 `results[0]` 的匹配内容，因为该对象不是伪数组。我们可以使用 `Array.from` 把它变成一个真正的 `Array`。在 Iterable（可迭代对象）<info:iterable>一文中有关于伪数组和可迭代对象的更多详细信息。

Reviewer 2: [我们无法获得 `results[0]` 的匹配内容，因为该对象不是伪数组] 
建议：[因为该对象是伪数组]
怀疑原文有误

原文：
because that object isn't pseudoarray

Author: 我觉得可能是[因为该对象不是数组]

如果我们不需要遍历结果，则 `Array.from` 没有必要：

```js run
let results = '<h1> <h2>'.matchAll(/<(.*?)>/gi);

for(let result of results) {
  alert(result);
  // 第一个结果: <h1>,h1
  // 第二个结果: <h2>,h2
}
```

> …或使用解构：

Reviewer: ```suggestion
……或使用解构：
```

alert( tag1.input ); // <h1> <h2>
```

> ```smart header="为什么是 `matchAll` 的结果是可迭代对象而不是数组？"

Reviewer 2: ```suggestion
```smart header="为什么 `matchAll` 的结果是可迭代对象而不是数组？"
```

> 为什么这个方法这样设计？原因很简单——为了优化。

Reviewer 2: ```suggestion
为什么这个方法这样设计？原因很简单 — 为了优化。
```

调用 `matchAll` 不会执行搜索。相反，它返回一个可迭代的对象，最初没有结果。每当我们对它进行迭代时才会执行搜索，例如在循环中。

因此，这将根据需要找到尽可能多的结果，而不是全部。

> 例如，文本中可能有100个匹配项，但是在一个 `for..of` 循环中，我们已经找到了 5 个匹配项，然后觉得足够了并做出一个 `break`。这时引擎就不会花时间查找其他 95 个匹配。

Reviewer 2: ```suggestion
例如，文本中可能有 100 个匹配项，但是在一个 `for..of` 循环中，我们已经找到了 5 个匹配项，然后觉得足够了并做出一个 `break`。这时引擎就不会花时间查找其他 95 个匹配。
```

// 30.10.2019, 01.01.2020
```

## 非捕获组 ?:

有时我们需要括号才能正确应用量词，但我们不希望它们的内容出现在结果中。

可以通过在开头添加 `pattern:?:` 来排除组。

例如，如果我们要查找 `pattern:(go)+`，但不希望括号内容（`go`）作为一个单独的数组项，则可以编写：`pattern:(?:go)+`。

在下面的示例中，我们仅将名称 `match:John` 作为匹配项的单独成员：

```js run
let str = "Gogogo John!";

*!*
// ?: 从捕获组中排除 'go'
let regexp = /(?:go)+ (\w+)/i;
*/!*

let result = str.match(regexp);

> alert( result[0] ); // Gogogo John (完全匹配)

Reviewer: ```suggestion
alert( result[0] ); // Gogogo John（完全匹配）
```

alert( result[1] ); // John

> alert( result.length ); // 2 (数组中没有更多项)

Reviewer: ```suggestion
alert( result.length ); // 2（数组中没有更多项）
```
